oprf/oprf.go

```diff
@@ -1,307 +1,248 @@
-// It implements:
-// For a Client:
-//   - Blind
-//   - Unblind
-//   - Finalize
+// Package oprf provides an Oblivious Pseudo-Random Function protocol.
+//
+// An Oblivious Pseudorandom Function (OPRFs) is a two-party protocol for
+// computing the output of a PRF. One party (the server) holds the PRF secret
+// key, and the other (the client) holds the PRF input.
+//
+// Obliviousness: Ensures that the server does not learn anything
+// about the client's input during the Evaluation step.
+//
+// Verifiability: Allows the client to verify that the server used
+// a committed secret key during Evaluation step.
+//
+// OPRF is defined on draft-irtf-cfrg-voprf: https://datatracker.ietf.org/doc/draft-irtf-cfrg-voprf
 //
-// For a Server:
-//   - Setup
-//   - Evaluate
-//   - VerifyFinalize
 package oprf
 
 import (
-	"crypto/sha256"
-	"crypto/sha512"
-	"crypto/subtle"
+	"crypto"
+	"crypto/rand"
 	"encoding/binary"
 	"errors"
-	"hash"
 	"io"
 
-	"github.com/cloudflare/circl/oprf/group"
+	"github.com/cloudflare/circl/group"
 )
 
-// SuiteID is a type that represents the ID of a Suite.
-type SuiteID uint16
+const (
+	version        = "VOPRF06-"
+	seedDST        = "Seed-"
+	challengeDST   = "Challenge-"
+	finalizeDST    = "Finalize-"
+	compositeDST   = "Composite-"
+	hashToGroupDST = "HashToGroup-"
+)
+
+// SuiteID identifies supported suites.
+type SuiteID = uint16
 
 const (
-	// OPRFP256 is the constant to represent the OPRF P-256 with SHA-512 (SSWU-RO) group.
+	// OPRFP256 represents the OPRF with P-256 and SHA-256.
 	OPRFP256 SuiteID = 0x0003
-	// OPRFP384 is the constant to represent the OPRF P-384 with SHA-512 (SSWU-RO) group.
+	// OPRFP384 represents the OPRF with P-384 and SHA-512.
 	OPRFP384 SuiteID = 0x0004
-	// OPRFP521 is the constant to represent the OPRF P-521 with SHA-512 (SSWU-RO) group.
+	// OPRFP521 represents the OPRF with P-521 and SHA-512.
 	OPRFP521 SuiteID = 0x0005
 )
 
-var (
-	// OPRFMode is the context string to define a OPRF.
-	OPRFMode byte = 0x00
-)
+// Mode specifies a variant of the OPRF protocol.
+type Mode = uint8
 
-var (
-	// ErrUnsupportedGroup is an error stating that the ciphersuite chosen is not supported
-	ErrUnsupportedGroup = errors.New("unsupported group")
+const (
+	BaseMode       Mode = 0x00
+	VerifiableMode Mode = 0x01
 )
 
-// BlindToken corresponds to a token that has been blinded.
-// Internally, it is a serialized Element.
-type BlindToken []byte
+// ErrUnsupportedSuite is thrown when requesting a non-supported suite.
+var ErrUnsupportedSuite = errors.New("non-supported suite")
 
-// IssuedToken corresponds to a token that has been issued.
-// Internally, it is a serialized Element.
-type IssuedToken []byte
+type Blind group.Scalar
+type Serialized = []byte
+type Blinded = []byte
 
-// Token is the object issuance of the protocol.
-type Token struct {
-	data  []byte
-	blind *group.Scalar
+type Proof struct {
+	PublicKey Serialized
```

Comment:
Why does the proof carry `PublicKey`?

Reply:
This might not be the better place. 
At which point the server's public key must be known by the client.? a) at client's inception  b) only during Finalize step.

Reply:
I opted by Option B.

Reply:
after Chris' review, we opted by option B.